thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/alert/AlertJobScheduler.java

```diff
@@ -66,23 +71,74 @@ public AlertJobScheduler(JobManager anomalyJobDAO, TaskManager anomalyTaskDAO,
+    scheduledExecutorService.scheduleWithFixedDelay(this, 0, 30, TimeUnit.MINUTES);
+  }
+
+  public void run() {
+    try {
+      // read all alert configs
+      LOG.info("Reading all alert configs..");
+      List<EmailConfigurationDTO> alertConfigs = emailConfigurationDAO.findAll();
+
+      // get active jobs
+      List<String> scheduledJobs = getScheduledJobs();
+      LOG.info("Scheduled jobs {}", scheduledJobs);
+
+      for (EmailConfigurationDTO alertConfig : alertConfigs) {
+        Long id = alertConfig.getId();
+        String jobKey = getJobKey(id);
+        boolean isActive = alertConfig.isActive();
+        boolean isScheduled = scheduledJobs.contains(jobKey);
+
+        // for all jobs with isActive, but not in scheduled jobs,
+        // schedule them with quartz, as function is newly created, or newly activated
+        if (isActive && !isScheduled) {
+          LOG.info("Found active but not scheduled {}", id);
+          startJob(alertConfig, jobKey);
+        }
+        // for all jobs with not isActive, but in scheduled jobs,
+        // remove them from quartz, as function is newly deactivated
+        else if (!isActive && isScheduled) {
+          LOG.info("Found inactive but scheduled {}", id);
+          stopJob(jobKey);
+        }
+        // for all jobs with isActive, and isScheduled,
+        // updates to a function will be picked up automatically by the next run
+        // but check for cron updates
+        else if (isActive && isScheduled) {
+          String cronInDatabase = alertConfig.getCron();
+          List<Trigger> triggers = (List<Trigger>) quartzScheduler.getTriggersOfJob(JobKey.jobKey(jobKey));
+          CronTrigger cronTrigger = (CronTrigger) triggers.get(0);
+          String cronInSchedule = cronTrigger.getCronExpression();
+          // cron expression has been updated, restart this job
+          if (!cronInDatabase.equals(cronInSchedule)) {
+            LOG.info("Cron expression for config {} with jobKey {} has been changed from {}  to {}. "
+                + "Restarting schedule", id, jobKey, cronInSchedule, cronInDatabase);
+            stopJob(jobKey);
+            startJob(alertConfig, jobKey);
+          }
+        }
+
+        // for all jobs with not isActive, and not isScheduled, no change required
       }
+
+      // for any scheduled jobs, not having a function in the database,
+      // stop the schedule, as function has been deleted
+      for (String scheduledJobKey : scheduledJobs) {
+        Long configId = getIdFromJobKey(scheduledJobKey);
+        EmailConfigurationDTO alertConfigSpec = emailConfigurationDAO.findById(configId);
+        if (alertConfigSpec == null) {
+          LOG.info("Found scheduled, but not in database {}", configId);
+          stopJob(scheduledJobKey);
+        }
+      }
+    } catch (SchedulerException e) {
+      LOG.error("Exception in reading active jobs", e);
     }
   }
 
   public void stop() throws SchedulerException {
```

Comment:
Could we rename stop() to shutdown(), because the behavior of shutdown() is already defined in executor service. On the other hand, stop() is not well defined and could mean anything such as 1. the thread pool is shutdown but threads are still running, 2. pool and threads are both terminated, etc.